File: src/CattyTests/Bricks/TurnLeftBrickTests.m
/**
 *  Copyright (C) 2010-2018 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import <XCTest/XCTest.h>
#import "AbstractBrickTests.h"
#import "WhenScript.h"
#import "SpriteObject.h"
#import "Pocket_Code-Swift.h"

@interface TurnLeftBrickTests : AbstractBrickTests
@end

@implementation TurnLeftBrickTests

- (void)setUp
{
    [super setUp];
    // Put setup code here; it will be run once, before the first test case.
}

- (void)tearDown
{
    // Put teardown code here; it will be run once, after the last test case.
    [super tearDown];
}

- (void)testTurnLeftBrick
{
    [self turnLeftWithInitialRotation:90 andRotation:60];
    [self turnLeftWithInitialRotation:0 andRotation:60];
    [self turnLeftWithInitialRotation:90 andRotation:400];
}

- (void)testTurnLeftBrickNegative
{
    [self turnLeftWithInitialRotation:90 andRotation:-60];
}

- (void)testTurnLeftBrickNegativeOver360
{
    [self turnLeftWithInitialRotation:90 andRotation:-400];
}

- (void)turnLeftWithInitialRotation:(CGFloat)initialRotation andRotation:(CGFloat)rotation
{
    rotation = fmodf(rotation, 360.0f);
    
    SpriteObject *object = [[SpriteObject alloc] init];
    CBSpriteNode *spriteNode = [[CBSpriteNode alloc] initWithSpriteObject:object];
    object.spriteNode = spriteNode;
    spriteNode.rotation = initialRotation;
    
    Script *script = [[WhenScript alloc] init];
    script.object = object;
    
    TurnLeftBrick* brick = [[TurnLeftBrick alloc] init];
    brick.script = script;
    
    Formula *degrees = [[Formula alloc] init];
    FormulaElement *formulaTree = [[FormulaElement alloc] init];
    formulaTree.type = NUMBER;
    formulaTree.value = [NSString stringWithFormat:@"%f", rotation];
    degrees.formulaTree = formulaTree;
    brick.degrees = degrees;
    
    dispatch_block_t action = [brick actionBlock];
    action();
    
    CGFloat expectedRotation = initialRotation - rotation;
    XCTAssertEqualWithAccuracy(expectedRotation, spriteNode.rotation, 0.0001, @"TurnLeftBrick not correct");
}

@end

File: src/CattyTests/Bricks/TurnLeftBrickTests.swift
/**
 *  Copyright (C) 2010-2022 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

import XCTest

@testable import Pocket_Code

final class TurnLeftBrickTests: AbstractBrickTest {

    func testTurnLeftBrick() {
        turnLeft(initialRotation: 90, rotation: 60)
        turnLeft(initialRotation: 0, rotation: 60)
        turnLeft(initialRotation: 90, rotation: 400)
    }

    func futestTurnLeftBrickNegative() {
        turnLeft(initialRotation: 90, rotation: -60)
    }

    func testTurnLeftBrickNegativeOver360() {
        turnLeft(initialRotation: 90, rotation: -400)
    }

    private func turnLeft(initialRotation: Double, rotation: Double) {
        let object = SpriteObject()
        let scene = Scene(name: "testScene")
        object.scene = scene
        object.name = "testname"
        let spriteNode = CBSpriteNode.init(spriteObject: object)
        object.spriteNode = spriteNode

        spriteNode.catrobatRotation = initialRotation

        let script = WhenScript()
        script.object = object

        let brick = TurnLeftBrick()
        brick.script = script

        brick.degrees = Formula(double: rotation)

        let action = brick.actionBlock(self.formulaInterpreter)
        action()

        let expectedRawRotation = RotationSensor.convertToRaw(userInput: initialRotation - rotation, for: object)
        XCTAssertEqual(CGFloat(expectedRawRotation), spriteNode.zRotation, accuracy: 0.0001, "TurnLeftBrick not correct")
    }

    func testGetFormulas() {
        let brick = TurnLeftBrick()
        brick.degrees = Formula(double: 1)
        var formulas = brick.getFormulas()

        XCTAssertEqual(formulas?.count, 1)
        XCTAssertEqual(brick.degrees, formulas?[0])

        brick.degrees = Formula(double: 22)
        formulas = brick.getFormulas()

        XCTAssertEqual(brick.degrees, formulas?[0])
    }

    func testMutableCopy() {
        let brick = TurnLeftBrick()
        brick.degrees = Formula(double: 2)

        let copiedBrick: TurnLeftBrick = brick.mutableCopy(with: CBMutableCopyContext(), andErrorReporting: true) as! TurnLeftBrick

        XCTAssertTrue(brick.isEqual(to: copiedBrick))
        XCTAssertFalse(brick === copiedBrick)
        XCTAssertTrue(brick.degrees.isEqual(to: copiedBrick.degrees))
        XCTAssertFalse(brick.degrees === copiedBrick.degrees)
    }
}
